internal/runtime/internal/controller/value_cache.go

```diff
@@ -1,88 +1,90 @@
 }
 
 // newValueCache creates a new ValueCache.
 func newValueCache() *valueCache {
 	return &valueCache{
-		components:      make(map[string]ComponentID),
-		args:            make(map[string]interface{}),
-		exports:         make(map[string]interface{}),
-		moduleArguments: make(map[string]any),
+		componentIds:    make(map[string]ComponentID, 0),
 		moduleExports:   make(map[string]any),
+		moduleArguments: make(map[string]any),
+		scope:           vm.NewScope(make(map[string]any)),
 	}
 }
 
-func (vc *valueCache) SetScope(scope *vm.Scope) {
-	vc.mut.Lock()
-	defer vc.mut.Unlock()
-	vc.scope = scope
-}
-
-// CacheArguments will cache the provided arguments by the given id. args may
-// be nil to store an empty object.
-func (vc *valueCache) CacheArguments(id ComponentID, args component.Arguments) {
+// UpdateScopeVariables updates the Variables map of the scope with a deep copy of the provided map.
+func (vc *valueCache) UpdateScopeVariables(variables map[string]any) {
+	if variables == nil {
+		return
+	}
 	vc.mut.Lock()
 	defer vc.mut.Unlock()
-
-	nodeID := id.String()
-	vc.components[nodeID] = id
-
-	var argsVal interface{} = make(map[string]interface{})
-	if args != nil {
-		argsVal = args
-	}
-	vc.args[nodeID] = argsVal
+	vc.scope.Variables = deepCopyMap(variables)
 }
 
 // CacheExports will cache the provided exports using the given id. exports may
 // be nil to store an empty object.
-func (vc *valueCache) CacheExports(id ComponentID, exports component.Exports) {
+func (vc *valueCache) CacheExports(id ComponentID, exports component.Exports) error {
 	vc.mut.Lock()
 	defer vc.mut.Unlock()
 
-	nodeID := id.String()
-	vc.components[nodeID] = id
+	variables := vc.scope.Variables
+	// Build nested maps.
+	for i := 0; i < len(id)-1; i++ {
+		t := id[i]
+		if _, ok := variables[t]; !ok {
+			variables[t] = make(map[string]any)
+		} else if _, ok := variables[t].(map[string]any); !ok {
+			return fmt.Errorf("expected a map but found a value for %q when trying to cache the export for %s", t, id.String())
+		}
+		variables = variables[t].(map[string]any)
+	}
 
-	var exportsVal interface{} = make(map[string]interface{})
+	var exportsVal any = make(map[string]any)
 	if exports != nil {
 		exportsVal = exports
 	}
-	vc.exports[nodeID] = exportsVal
+	variables[id[len(id)-1]] = exportsVal
+	return nil
+}
+
+func (vc *valueCache) GetModuleArgument(key string) (any, bool) {
+	vc.mut.RLock()
+	defer vc.mut.RUnlock()
+	v, exist := vc.moduleArguments[key]
+	return v, exist
 }
 
 // CacheModuleArgument will cache the provided exports using the given id.
 func (vc *valueCache) CacheModuleArgument(key string, value any) {
 	vc.mut.Lock()
 	defer vc.mut.Unlock()
 
-	if value == nil {
-		vc.moduleArguments[key] = nil
-	} else {
-		vc.moduleArguments[key] = value
-	}
+	keyMap := make(map[string]any)
+	keyMap["value"] = value
+	vc.moduleArguments[key] = keyMap
```

Comment:
because to refer to an argument value in the config file you need to write "argument.LABEL.value". To resolve it, it must have the path ["argument"][LABEL]["value"]

Reply:
Ah gotcha. Yeah now I remember this syntax. Would things break if we made both `foo.bar.value` and `foo.bar` work? It could be UX improvement 🤔

Reply:
that would be nice but it's tricky to do because you cant have two values for the same key in the map + you would need to handle the weird cases where the user names its argument "value".